### DISCUSSION TITLE
Router does not send unused subscription variables to subgraphs

### DISCUSSION THREAD

**Discussion by vadimofd** [wundergraph/cosmo#700] [Q&A] [answered]
It looks like router does not send unused subscription variables to subgraphs.

Is there any documentation on this? Or could someone help me understand the need for this behavior?

As can be seen in the logs below the client sends an extra unused variable `test_var2` to the router (see last line of the client log).

However, the router does not send it to the subgraph (see last line of the subgraph log).

Client Log:
```
Connected (press CTRL+C to quit)
> {"type":"connection_init"}
< {"type":"connection_ack"}
> {"type":"start","id":"1","payload":{"variables":{"test_var":"test_val","test_var2":"test_val2"},"query":"subscription($test_var:String) {testWithVars(test:$test_var) {__typename}}"}}
```

Subgraph Log:
```
Client connected
< {"type":"connection_init","payload":{"extensions":{"upgradeHeaders":{"X-Forwarded-For":[REDUCTED],"X-Forwarded-Proto":["https"]}}}}
> {"type":"connection_ack"}
< {"type":"start","id":"1","payload":{"query":"subscription($test_var: String){testWithVars(test: $test_var){__typename}}","variables":{"test_var":"test_val"},"extensions":{"upgradeHeaders":{"X-Forwarded-For":[REDUCTED],"X-Forwarded-Proto":["https"]}}}}
```

Schema:
```GraphQL
schema {
    query: Query
    subscription: Subscriptions
}

type Query {
    dummy: String
}

type Subscriptions {
    testWithVars(test: String): String
}
```

**Comment by StarpTech** [comment]
HI @vadimofd, why do you want to send unused variables to the subgraph?

**Reply by vadimofd** [reply]
Hi, our existing GraphQL server was using these variables to receive some authentication details from the client when initiating a WebSocket subscription. With the router in the middle this mechanism broke and it was difficult to figure out the reason due to lack of router subscriptions tracing and this undocumented behavior.

**Reply by StarpTech** [reply, marked as answer]
Our engine normalize the query and removes unused variables. I recommend to define a proper contract in your schema or use headers to pass authentication information. The router can forward those headers to the subgraph.

**Reply by vadimofd** [reply]
Thank you. Would you consider adding this information and your recommendations to Cosmo documentation? It might also be helpful to log this happening in trace/debug logs and possibly even warn about in the Playground UI.

**Reply by StarpTech** [reply]
We consider this expected behavior. It is akin to an inner query in SQL where a field is requested but not utilized in the outer query. For efficiency, it is removed after normalization. We have not encountered this scenario frequently enough to justify adding logs for it. I hope you understand. Thank you!
